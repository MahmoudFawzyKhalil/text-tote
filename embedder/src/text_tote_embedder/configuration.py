import os
from dataclasses import dataclass
from os import path

from decouple import AutoConfig

local_dev_settings_or_test_settings = path.abspath('./settings.ini')
user_home = path.expanduser('~/texttote')
config_root: str = './' if path.exists(local_dev_settings_or_test_settings) else user_home
os.makedirs(config_root, exist_ok=True)
config = AutoConfig(search_path=config_root)


@dataclass
class AppConfig:
    db_path: str


db_path = path.abspath(path.expanduser(config('DB_PATH', default='~/texttote/texttote.db')))

appconfig = AppConfig(
    db_path=db_path
)
