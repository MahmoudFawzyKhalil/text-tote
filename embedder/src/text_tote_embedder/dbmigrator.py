import os
import re
import sqlite3


def execute_sql_file(cursor, file_path: str):
    with open(file_path, 'r') as sql_file:
        print(f"Executing SQL script: {file_path}")
        sql_script = sql_file.read()
        cursor.executescript(sql_script)


def extract_version(file_name: str) -> int:
    match = re.match(r'V(\d+)', file_name)
    if match:
        return int(match.group(1))
    else:
        raise Exception(f"Invalid file name: {file_name}")


def create_schema_version_table(cursor):
    # Create the schema_version table if it doesn't exist
    cursor.execute("CREATE TABLE IF NOT EXISTS schema_version (version integer PRIMARY KEY)")
    # check if we already have a row
    cursor.execute("SELECT * FROM schema_version")
    if cursor.fetchone():
        return
    else:
        cursor.execute("INSERT INTO schema_version (version) VALUES (0)")


def get_current_schema_version(cursor):
    # Fetch the current schema version from the database or return a default value if the table is not present
    cursor.execute("SELECT version FROM schema_version")
    result = cursor.fetchone()
    return result[0] if result else 0


def update_schema_version(cursor, version):
    # Update the schema_version table with the new version
    cursor.execute("UPDATE schema_version SET version = ?", (version,))


def execute_migrations(db_connection: sqlite3.Connection):
    cursor = db_connection.cursor()

    try:
        script_directory = os.path.dirname(os.path.abspath(__file__))
        migrations_directory = os.path.join(script_directory, 'migrations')

        sql_scripts = sorted(os.listdir(migrations_directory))

        # Create the schema_version table if it doesn't exist
        create_schema_version_table(cursor)

        # Fetch the current schema version
        current_version = get_current_schema_version(cursor)
        print(f"Current Schema Version: {current_version}")

        for sql_script in sql_scripts:
            if sql_script.endswith('.sql'):
                file_path = os.path.join(migrations_directory, sql_script)

                version = extract_version(sql_script)
                if version > current_version:
                    execute_sql_file(cursor, file_path)
                    update_schema_version(cursor, version)
                    print(f"Schema version updated to: {version}")

        db_connection.commit()
        print("SQL scripts executed successfully.")

    except Exception as e:
        print(f"Error executing SQL scripts: {e}")

    finally:
        db_connection.close()
