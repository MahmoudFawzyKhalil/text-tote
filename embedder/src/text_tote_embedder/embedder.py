from sentence_transformers import SentenceTransformer
from torch import Tensor

# https://www.sbert.net/examples/applications/semantic-search/README.html#symmetric-vs-asymmetric-semantic-search
# Model chosen because our queries are asymmetric queries
MODEL_NAME = 'msmarco-MiniLM-L-6-v3'
VECTOR_SIZE = 384

model = None


def lazy_load_model() -> SentenceTransformer:
    global model
    if not model:
        model = SentenceTransformer(MODEL_NAME)
    return model


def embed(sentences: str | list[str], show_progress_bar: bool = False) -> list[Tensor]:
    return lazy_load_model().encode(sentences, show_progress_bar=show_progress_bar)
