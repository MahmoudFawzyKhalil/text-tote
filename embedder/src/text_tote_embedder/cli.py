from .configuration import appconfig

def app():
    print(appconfig)
