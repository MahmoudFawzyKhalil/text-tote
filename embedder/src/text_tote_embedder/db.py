import os
import sqlite3

import sqlite_vss

from . import dbmigrator
from .configuration import appconfig


def create_db():
    if os.path.exists(appconfig.db_path):
        return

    uri = f"file:{appconfig.db_path}?mode=rwc"
    conn = sqlite3.connect(uri, uri=True)
    load_extensions(conn)


def migrate_db():
    db_connection = create_connection()
    dbmigrator.execute_migrations(db_connection)


def create_connection() -> sqlite3.Connection:
    uri = f"file:{appconfig.db_path}?mode=rw"
    conn = sqlite3.connect(uri, uri=True)
    load_extensions(conn)
    return conn


def load_extensions(conn: sqlite3.Connection):
    conn.enable_load_extension(True)
    sqlite_vss.load(conn)
