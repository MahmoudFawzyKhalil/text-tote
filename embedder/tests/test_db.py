import os

import approvaltests

from src.text_tote_embedder import db
from src.text_tote_embedder.configuration import appconfig


def test_migrate_db():
    try:
        # Arrange
        db.create_db()

        # Act
        db.migrate_db()

        # Assert
        db_connection = db.create_connection()
        metadata = get_table_metadata(db_connection)
        approvaltests.verify_as_json(metadata)
    finally:
        # Cleanup
        # os.remove(appconfig.db_path)
        pass


def get_table_metadata(connection):
    cursor = connection.cursor()

    # Get a list of all tables in the database
    cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
    tables = cursor.fetchall()

    # Dictionary to store table metadata
    table_metadata = {}

    # Loop through each table and retrieve metadata
    for table in tables:
        table_name = table[0]
        table_metadata[table_name] = {}

        # Get table definition and metadata
        cursor.execute(f"PRAGMA table_info({table_name});")
        columns = cursor.fetchall()

        # Store column names and definitions in the dictionary
        table_metadata[table_name]['columns'] = []
        for column in columns:
            column_name = column[1]
            column_type = column[2]
            is_primary_key = column[5]
            table_metadata[table_name]['columns'].append({
                'name': column_name,
                'type': column_type,
                'is_primary_key': bool(is_primary_key)
            })

    return table_metadata
