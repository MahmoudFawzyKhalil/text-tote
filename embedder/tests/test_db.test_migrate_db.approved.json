{
    "resource": {
        "columns": [
            {
                "is_primary_key": true,
                "name": "id",
                "type": "INTEGER"
            },
            {
                "is_primary_key": false,
                "name": "url",
                "type": "TEXT"
            },
            {
                "is_primary_key": false,
                "name": "title",
                "type": "TEXT"
            }
        ]
    },
    "schema_version": {
        "columns": [
            {
                "is_primary_key": true,
                "name": "version",
                "type": "INTEGER"
            }
        ]
    }
}
